packages/flutter/lib/src/widgets/splash.dart

```diff
@@ -0,0 +1,621 @@
+  /// use this vsync to drive their animations.
+  TickerProvider get vsync;
+
+  /// Adds an [InkFeature], such as an [InkSplash] or an [InkHighlight].
+  ///
+  /// The ink feature will paint as part of this controller.
+  @Deprecated(
+    'Use addSplash instead. '
+    '"Splash effects" no longer rely on a MaterialInkController. '
+    'This feature was deprecated after v3.23.0-0.1.pre.',
+  )
+  void addInkFeature(Splash feature);
+
+  /// Adds a [Splash], such as an [InkSplash] or an [InkHighlight].
+  ///
+  /// The splash will paint as part of this controller.
+  void addSplash(Splash splash);
+
+  /// Removes a [Splash] added by [addSplash].
+  void removeSplash(Splash splash);
+
+  /// A function called when the controller's layout changes.
+  ///
+  /// [RenderBox.markNeedsPaint] should be called if there are
+  /// any active [Splash]es.
+  void didChangeLayout();
+}
+
+/// {@template flutter.widgets.splash.SplashBox}
+/// Defines an area for descendant [Splash]es to paint on.
+///
+/// Used by the [Material] widget to enable button ink effects.
+///
+/// There are a few reasons that using a `SplashBox` directly might be
+/// preferred over a `Material`:
+///
+/// * A [Decoration] can be added without the [downsides of the `Ink` widget](https://api.flutter.dev/flutter/widgets/Ink-class.html#limitations).
+/// * `SplashBox` doesn't use [implicit animations](https://docs.flutter.dev/codelabs/implicit-animations),
+///   offering more granular control over UI properties.
+///   (This is especially helpful when its properties come from values that
+///   are already being animated, such as when `Theme.of(context).colorScheme`
+///   inherits from a `MaterialApp`'s `AnimatedTheme`.)
+/// * If a Flutter app isn't using the [Material design system](https://m3.material.io/),
+///   `SplashBox` is the easiest way to add [Splash]es.
+/// {@endtemplate}
+///
+/// Splashes are painted using a [SplashController], a render object that
+/// re-paints each [Splash] when the widget's layout changes.
+///
+/// {@tool snippet}
+/// Generally, a [SplashBox] should be set as the child of widgets that
+/// perform clipping and decoration, and it should be the parent of widgets
+/// that create [Splash]es.
+///
+/// Example:
+///
+/// ```dart
+/// ClipRRect(
+///   borderRadius: BorderRadius.circular(8),
+///   child: ColoredBox(
+///     color: color,
+///     child: const SplashBox(
+///       // add an InkWell here,
+///       // or a different child that creates Splash effects
+///     ),
+///   ),
+/// );
+/// ```
+/// {@end-tool}
+///
+/// {@tool dartpad}
+/// This example shows how to make a button using a [SplashBox].
+///
+/// ** See code in examples/api/lib/widgets/splash_box/splash_box.0.dart **
+/// {@end-tool}
+///
+/// See also:
+/// * [SplashController], used by this widget to enable splash effects.
+/// * [Splash], the class that holds splash effect data.
+class SplashBox extends StatefulWidget {
```

Comment:
I kind of prefer the name `SplashResponse`, which corresponds to `InkResponse`. If I understand correctly, `InkResponse` does not assume the box protocol, and neither should this class.

Reply:
Oh! Actually, I think I might be able to help improve your understanding here.

Currently, this pull request doesn't have a class that corresponds to `InkResponse`. The current plan is to add a `SplashResponse` or something similar in a future PR: it would essentially be an `InkResponse` that doesn't use any theme data or splash factories by default, and then `InkResponse` would probably extend off of it.

<br>

`_InkFeatures` is a private class that was originally in the same library as the `Material` widget.
Eventually, I'll probably change the private class names to something without "ink", but surprisingly, doing this would break many different tests, since they're set up to detect the private class name using `runtimeType.toString()`.

`SplashBox` is just a wrapper for `_InkFeatures` that adds a `NotificationListener`. This widget uses the box protocol by default, but clipping can be applied to turn it into any shape.